src/intellisense/texDocSymbolProvider.ts

```diff
@@ -0,0 +1,314 @@
+    const filePath = document.fileName;
+    if (filePath === undefined) {
+        return [];
+    }
+    const rootElement = { children: [] };
+    let ast = await unifiedParser.parse(document.getText());
+    for (const node of ast.content) {
+        if (['string', 'parbreak', 'whitespace'].includes(node.type)) {
+            continue;
+        }
+        try {
+            await parseNode(node, rootElement);
+        }
+        catch (e) {
+            console.log(e);
+        }
+    }
+    let struct = rootElement.children as TeXElement[];
+    struct = hierarchyStructFormat(struct);
+    return struct;
+}
+
+/*
+    * Format the tree-like TeXElement[] structure based on the given order in defaultStructure
+    * 
+    * @param struct: the flat TeXElement[] structure
+    * @return: the formatted TeXElement[] structure
+*/
+function hierarchyStructFormat(struct: TeXElement[]): TeXElement[] {
+    const newStruct: TeXElement[] = [];
+    const orderFromDefaultStructure = new Map<string, number>();
+    for (const [i, section] of defaultStructure.entries()) {
+        orderFromDefaultStructure.set(section, i);
+    }
+    let prev: TeXElement | undefined;
+    for (const section of struct) {
+        // Root Node
+        if (prev === undefined) {
+            newStruct.push(section);
+            prev = section;
+            continue;
+        }
+        // Comparing the order of current section and previous section from defaultStructure
+        if ((orderFromDefaultStructure.get(section.name) ?? defaultStructure.length) > (orderFromDefaultStructure.get(prev.name) ?? defaultStructure.length)) {
+            prev.children.push(section);
+        } else {
+            newStruct.push(section);
+            prev.children = hierarchyStructFormat(prev.children);
+            prev = section;
+        }
+    }
+    if (prev !== undefined) {
+        prev.children = hierarchyStructFormat(prev.children);
+    }
+    return newStruct;
+}
+
+
+
+/*
+    * Parse a node and generate a TeXElement, this function travel each node recursively (Here children is referenced by `.content`),
+    * if the node is a macro and matches the defaultStructure, it will be treated as a section;
+    * if the node is a macro and matches the defaultCMD, it will be treated as a command;
+    * if the node is an environment, it will be treated as an environment;
+    * 
+    * @param node: the node to be parsed
+    * @param root: the root TeXElement
+*/
+// reference: https://github.com/James-Yu/LaTeX-Workshop/blob/master/src/outline/structurelib/latex.ts#L86
+async function parseNode(
+        node: Ast.Node,
+        root: { children: TeXElement[] }) {
+    const attributes = {
+        lineFr: (node.position?.start.line ?? 1) - 1,
+        lineTo: (node.position?.end.line ?? 1) - 1,
+        children: []
+    };
+    let element: TeXElement | undefined;
+    if (node.type === 'macro'&& defaultStructure.includes(node.content) && node.args?.[3].openMark === '{') {
```

Comment:
use `switch-case` syntax. And could you please improve the readability?

Reply:
By using switch and case, and some if-else statements are transformed into forms like `case 'section'` which might improve the readability a little bit.